### PR TITLE
bpo-45774: Autoconfiscate SQLite detection (GH-29507)

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue45774

### CODE REVIEW

**Comment by erlend-aasland** [comment]
SQLite was not detected on buildbot `AMD64 FreeBSD Shared PR`.

**Comment by tiran** [comment]
> SQLite was not detected on buildbot `AMD64 FreeBSD Shared PR`.

@koobs where is sqlite installed on your FreeBSD buildbot? Is it in /usr/local by any chance? Would the solution https://bugs.python.org/issue45774#msg406076 work for FreeBSD?

**Comment by koobs** [comment]
> > SQLite was not detected on buildbot `AMD64 FreeBSD Shared PR`.
> 
> @koobs where is sqlite installed on your FreeBSD buildbot? Is it in /usr/local by any chance? Would the solution https://bugs.python.org/issue45774#msg406076 work for FreeBSD?

sqlite3-3.35.5 is installed on both FreeBSD buildbot workers in LOCALBASE (`/usr/local` by default). This is the standard location for third party libraries installed via FreeBSD ports and official packages.

I haven't deeply reviewed the implementation in issue 45774, but I don't see a reference to `--with(out)-sqlite3` to explicitly disable/enable, but otherwise, generally speaking, standard configure checks for dependencies is fine.

I would also caution that wherever possible when doing dependency detection and configuration, we should include the ability to pass dependency specific include/library paths and *FLAGS, because Pythons build system is particularly inconsistent and flaky with respect to passing *FLAGS globally, which downstreams often have to utilise. 

This can causes include ordering issues, as well as build/link time conflicts when a dependency is installed in more than one location, and both are picked up.

Other than at the configure level `--with-foo-{inc,lib}=<path> (or similar)`, this also means `Modules/Setup`, if support for it is not there already (like SSL). 

## Further Sqlite3 Information

Installed libraries and header names:

```
# ls -la /usr/local/lib/libsqlite3.so*
lrwxr-xr-x  1 root  wheel       19 Oct 30 01:09 /usr/local/lib/libsqlite3.so@ -> libsqlite3.so.0.8.6
lrwxr-xr-x  1 root  wheel       19 Oct 30 01:09 /usr/local/lib/libsqlite3.so.0@ -> libsqlite3.so.0.8.6
-rwxr-xr-x  1 root  wheel  1422160 Oct 30 01:09 /usr/local/lib/libsqlite3.so.0.8.6*

# ls -la /usr/local/include/sql*
-rw-r--r--  1 root  wheel     838 Oct  4 08:15 /usr/local/include/sql3types.h
-rw-r--r--  1 root  wheel    1283 Oct  4 08:15 /usr/local/include/sqlca.h
-rw-r--r--  1 root  wheel    1586 Oct  4 08:15 /usr/local/include/sqlda-compat.h
-rw-r--r--  1 root  wheel     824 Oct  4 08:15 /usr/local/include/sqlda-native.h
-rw-r--r--  1 root  wheel     321 Oct  4 08:15 /usr/local/include/sqlda.h
-rw-r--r--  1 root  wheel  584223 Oct 30 01:09 /usr/local/include/sqlite3.h
-rw-r--r--  1 root  wheel   35437 Oct 30 01:09 /usr/local/include/sqlite3ext.h

sqlite also installed a pkg-config file:

```
# pkgconf --list-all |grep sql
# sqlite3                        SQLite - SQL database engine

# pkgconf sqlite3 --cflags --libs
# -I/usr/local/include -L/usr/local/lib -lsqlite3
```

**Comment by tiran** [comment]
Thanks @koobs !

Yes, I agree. ``setup.py`` does some problematic things in regards of default include and linker path. We are working on a series of improvements for the build system. One goal is to move header and linker checks out of ``setup.py`` and improve ``Modules/Setup.local``.

We don't use pkg-config for sqlite3 yet. Is ``/usr/local/include`` on the default search path?

**Comment by koobs** [comment]
We should assume it is (and not hardcode paths, unless there's no better alternative). Right now most if not all downstreams will be passing -isystem and/or custom build *FLAGS (along with CC et al if needed) at the global environment, and/or configure environment, and/or make environment and or Modules/Setup level to make the build work, given the current state of the Python build system. Adding another unique or special/different case would be another step backward IMO.

Sorry for the non-binary answer :)

**Comment by tiran** [comment]
FYI, I summarized my plan in https://bugs.python.org/issue45573#msg406084

**Comment by erlend-aasland** [comment]
FYI: Holding this until GH-29164 has landed.

**Comment by erlend-aasland** [comment]
SQLite is now detected by pkg-config on FreeBSD (builds fine as well). Autoconf complains, though, but it does not seem to matter much:

```
checking sqlite3.h usability... yes
checking sqlite3.h presence... no
configure: WARNING: sqlite3.h: accepted by the compiler, rejected by the preprocessor!
configure: WARNING: sqlite3.h: proceeding with the compiler's result
checking for sqlite3.h... yes
checking for sqlite3_open_v2 in -lsqlite3... yes
checking for sqlite3_load_extension in -lsqlite3... yes
checking for --enable-loadable-sqlite-extensions... no 
```

**Comment by erlend-aasland** [comment]
@tiran can you have a look? When it comes to `Modules/Setup`, I'm not quite comfortable.

File: Makefile.pre.in [hunk 1]
```diff
@@ -564,6 +564,14 @@ LIBEXPAT_HEADERS= \
 
 LIBEXPAT_A= Modules/expat/libexpat.a
 
+#
+# SQLite stuff
+#
+
+# For setup.py
+PY_SQLITE_CFLAGS=@PY_SQLITE_CFLAGS@
+PY_SQLITE_LDFLAGS=@PY_SQLITE_LDFLAGS@
```
Comment (tiran) [inline Makefile.pre.in]: My PR GH-29534 adds some automated variables like ``MODULE_{name}_CFLAGS``. Let me pull out the interesting bits as a separate PR so you can build on top of it.

```
MODULE__SQLITE3_=yes  # yes, missing, disabled, n/a
MODULE__SQLITE3_CFLAGS=
MODULE_SQLITE3_LDFLAGS=
```

File: Modules/Setup [hunk 2]
```diff
@@ -218,7 +218,7 @@ time timemodule.c
 #_dbm _dbmmodule.c -lgdbm_compat -DUSE_GDBM_COMPAT
 #_gdbm _gdbmmodule.c -lgdbm
 #_lzma _lzmamodule.c -llzma
-#_sqlite3 _sqlite/connection.c _sqlite/cursor.c _sqlite/microprotocols.c _sqlite/module.c _sqlite/prepare_protocol.c _sqlite/row.c _sqlite/statement.c _sqlite/util.c -lsqlite3
+#_sqlite3 _sqlite/connection.c _sqlite/cursor.c _sqlite/microprotocols.c _sqlite/module.c _sqlite/prepare_protocol.c _sqlite/row.c _sqlite/statement.c _sqlite/util.c $(PY_SQLITE_LDFLAGS)
```
Comment (tiran) [inline Modules/Setup]: Please include CFLAGS here, too.

File: configure.ac [hunk 3]
```diff
@@ -3165,12 +3165,48 @@ if test "$have_glibc_memmove_bug" = yes; then
     AS_VAR_APPEND([LIBMPDEC_CFLAGS], [" -U_FORTIFY_SOURCE"])
 fi
 
+dnl Check for SQLite library. Use pkg-config if available.
+AC_SUBST([PY_SQLITE_CFLAGS])
+AC_SUBST([PY_SQLITE_LDFLAGS])
+if test ! -z "$PKG_CONFIG"; then
+  AS_VAR_APPEND([PY_SQLITE_CFLAGS], ["`"$PKG_CONFIG" sqlite3 --cflags-only-I 2>/dev/null`"])
+  AS_VAR_APPEND([PY_SQLITE_LDFLAGS], ["`"$PKG_CONFIG" sqlite3 --libs-only-L 2>/dev/null`"])
```
Comment (tiran) [inline configure.ac]: I recommend to use full cflags and ldflags here. I also have some code that handles other flags.

File: Makefile.pre.in (hunk 1 above)
Comment (erlend-aasland) [reply Makefile.pre.in]: SGTM!

File: Modules/Setup (hunk 2 above)
Comment (erlend-aasland) [reply Modules/Setup]: Hm, I thought I already did, but I guess that bit disappeared in a `git reset` somewhere before last commit :) Will fix.

File: configure.ac (hunk 3 above)
Comment (erlend-aasland) [reply configure.ac]: SGTM

File: Makefile.pre.in (hunk 1 above)
Comment (erlend-aasland) [reply Makefile.pre.in]: Leaving this PR as draft until you've merged your new PR.

File: Makefile.pre.in [hunk 4]
```diff
@@ -2481,6 +2481,7 @@ MODULE__SHA512_DEPS=$(srcdir)/Modules/hashlib.h
 MODULE__SOCKET_DEPS=$(srcdir)/Modules/socketmodule.h
 MODULE__SSL_DEPS=$(srcdir)/Modules/_ssl.h $(srcdir)/Modules/_ssl/cert.c $(srcdir)/Modules/_ssl/debughelpers.c $(srcdir)/Modules/_ssl/misc.c $(srcdir)/Modules/_ssl_data.h $(srcdir)/Modules/_ssl_data_111.h $(srcdir)/Modules/_ssl_data_300.h $(srcdir)/Modules/socketmodule.h
 MODULE__TESTCAPI_DEPS=$(srcdir)/Modules/testcapi_long.h
+MODULE__SQLITE3_DEPS=$(srcdir)/Modules/_sqlite/connection.h $(srcdir)/Modules/_sqlite/cursor.h $(srcdir)/Modules/_sqlite/microprotocols.h $(srcdir)/Modules/_sqlite/module.h $(srcdir)/Modules/_sqlite/prepare_protocol.h $(srcdir)/Modules/_sqlite/row.h $(srcdir)/Modules/_sqlite/util.h
```
Comment (erlend-aasland) [inline Makefile.pre.in:2484]: Hm, this is left from some experiment I did last week or so. AFAICS, it's not needed, so I guess it can go.

File: configure.ac [hunk 5]
```diff
@@ -5999,6 +6032,10 @@ PY_STDLIB_MOD([pyexpat], [], [], [$LIBEXPAT_CFLAGS], [$LIBEXPAT_LDFLAGS])
 PY_STDLIB_MOD([_elementtree], [], [], [$LIBEXPAT_CFLAGS], [])
 
 PY_STDLIB_MOD([_decimal], [], [], [$LIBMPDEC_CFLAGS], [$LIBMPDEC_LDFLAGS])
+PY_STDLIB_MOD([_sqlite3],
+  [test "$have_sqlite3" = "yes"],
+  [test "$have_supported_sqlite3" = "yes"],
```
Comment (erlend-aasland) [inline configure.ac:6044]: - SQLite not present => disabled
- SQLite present, version not ok => missing
- SQLite present, version ok => yes

**Comment by erlend-aasland** [comment]
`pkg-config` is now used, if available. Fallback to default `autoconf` paths if `pkg-config` is not available, or if `pkg-config` cannot find a SQLite library with the required version.

Possible improvement: invalidate cache when `--pkg-config` status change. I'm -0.

**Comment by erlend-aasland** [comment]
If this looks good to you, @tiran, we can run it through the build bots once more, to verify that FreeBSD still works, but I'm pretty sure it does; the last buildbot run had (rude) pkg-config support, and SQLite was detected.

File: Modules/Setup [hunk 6]
```diff
@@ -227,7 +227,7 @@ time timemodule.c
 #_dbm _dbmmodule.c -lgdbm_compat -DUSE_GDBM_COMPAT
 #_gdbm _gdbmmodule.c -lgdbm
 #_lzma _lzmamodule.c -llzma
-#_sqlite3 _sqlite/connection.c _sqlite/cursor.c _sqlite/microprotocols.c _sqlite/module.c _sqlite/prepare_protocol.c _sqlite/row.c _sqlite/statement.c _sqlite/util.c -lsqlite3
+#_sqlite3 _sqlite/connection.c _sqlite/cursor.c _sqlite/microprotocols.c _sqlite/module.c _sqlite/prepare_protocol.c _sqlite/row.c _sqlite/statement.c _sqlite/util.c
```
Comment (erlend-aasland) [inline Modules/Setup]: Perhaps I should pull in `main` and add a template to `Modules/Setup.stdlib.in`, now that GH-29615 is merged?

File: configure.ac [hunk 7]
```diff
@@ -3165,12 +3165,45 @@ if test "$have_glibc_memmove_bug" = yes; then
     AS_VAR_APPEND([LIBMPDEC_CFLAGS], [" -U_FORTIFY_SOURCE"])
 fi
 
+dnl Check for SQLite library. Use pkg-config if available.
+PKG_CHECK_MODULES(
```
Comment (erlend-aasland) [inline configure.ac:3169]: @tiran: this macro is pretty neat 😎

File: configure.ac [hunk 8]
```diff
@@ -3165,12 +3165,45 @@ if test "$have_glibc_memmove_bug" = yes; then
...
+  [LIBSQLITE3], [sqlite3 >= 3.7.15], [], [
```
Comment (tiran) [inline configure.ac:3170]: Oh nice, I was not aware that ``MODULES`` are even takes a version. Nice!

File: configure.ac [hunk 9]
```diff
@@ -3165,12 +3165,45 @@ if test "$have_glibc_memmove_bug" = yes; then
...
+    LIBSQLITE3_LIBS="-lsqlite3"
+    LIBSQLITE3_CFLAGS=
+  ]
+)
+AS_VAR_APPEND([LIBSQLITE3_CFLAGS], [' -I$(srcdir)/Modules/_sqlite'])
```
Comment (tiran) [inline configure.ac:3175]: Do single quotes perform quoting of ``$`` symbol? I used ``\$(srcdir)`` in other places. Single quotes are nicer, though.

File: configure.ac [hunk 10]
```diff
@@ -3165,12 +3165,45 @@ if test "$have_glibc_memmove_bug" = yes; then
...
+
+AS_VAR_COPY([save_CFLAGS], [CFLAGS])
+AS_VAR_COPY([save_LDFLAGS], [LDFLAGS])
+AS_VAR_COPY([CFLAGS], [LIBSQLITE3_CFLAGS])
+AS_VAR_COPY([LDFLAGS], [LIBSQLITE3_LIBS])
```
Comment (tiran) [inline configure.ac]: Idea for another PR: We push and pop LIBS, CFLAGS, and LDFLAGS pretty often. How do you feel about two macros to automate the task?

File: configure.ac [hunk 11]
```diff
@@ -3165,12 +3165,45 @@ if test "$have_glibc_memmove_bug" = yes; then
...
+AS_VAR_COPY([LDFLAGS], [save_LDFLAGS])
+
 # Check for support for loadable sqlite extensions
 AC_MSG_CHECKING(for --enable-loadable-sqlite-extensions)
 AC_ARG_ENABLE(loadable-sqlite-extensions,
               AS_HELP_STRING([--enable-loadable-sqlite-extensions],
                              [support loadable extensions in _sqlite module, see Doc/library/sqlite3.rst (default is no)]),
-              [],
+              [AS_VAR_IF([have_sqlite3_load_extension], [no],
+                         [AC_MSG_ERROR([Your version of SQLite does not support loadable extensions])])],
```
Comment (tiran) [inline configure.ac]:
```suggestion
                         [AC_MSG_WARN([Your version of SQLite does not support loadable extensions])])],
```

``AC_MSG_ERROR`` aborts configure. Is that a good idea? We do not abort configure when an optional feature is missing.

File: configure.ac (hunk 9 above)
Comment (erlend-aasland) [reply configure.ac:3175]: Yeah, I prefer single quotes to escape escaping :)

File: configure.ac (hunk 10 above)
Comment (erlend-aasland) [reply configure.ac]: Yes! I've been looking for something like that in the autoconf docs, and it surprises me that it does not exist. It seems to be a very common operation. I think I would prefer a push/pop kind of naming.

File: configure.ac (hunk 11 above)
Comment (erlend-aasland) [reply configure.ac]: Oops, that's true. No, we should definitely continue. Thanks for catching that.

File: configure.ac (hunk 9 above)
Comment (tiran) [reply configure.ac:3175]: Yeah, me too. I was not aware that single quotes work here. I'll update my other code later.

Comment (ned-deily) [reply configure.ac:3175]: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html  Section 2.2 Quoting

**Comment by tiran** [comment]
Looks good! Great job! Let's do another buildbot run. I'll merge the PR tomorrow.

**Comment by erlend-aasland** [comment]
FreeBSD still emits a warning:

```
checking sqlite3.h usability... yes
checking sqlite3.h presence... no
configure: WARNING: sqlite3.h: accepted by the compiler, rejected by the preprocessor!
configure: WARNING: sqlite3.h: proceeding with the compiler's result
checking for sqlite3.h... yes
checking for sqlite3_open_v2 in -lsqlite3... yes
checking for sqlite3_load_extension in -lsqlite3... yes
```

It also failed the configure step, but that was bco. `getaddrinfo`:
```
Fatal: You must get working getaddrinfo() function.
       or you can specify "--disable-ipv6".
```

**Comment by erlend-aasland** [comment]
The ARM64 macOS buildbot failed at `AC_CHECK_LIB([sqlite3], [sqlite3_open_v2], ...`, leaving `sqlite3` disabled. IIRC, macOS 11 had SQLite 3.23.x installed; there should be no reason for it to fail.

**Comment by tiran** [comment]
You figured it out already!

FreeBSD has 3rd party headers in /usr/local/include. The pre-processor does not use CFLAGS. You need CPPFLAGS here. I concur that it's safe to use CPPFLAGS instead of CFLAGS for sqlite. Most pgkconf files set -I, -D, -L, and -l. Other rare cases may need both CFLAGS and CPPFLAGS, though.

**Comment by erlend-aasland** [comment]
If 41a78b5b36d9c82ba6ce3bd00688bba32be55ba4 looks ok for you (as a fix for both the FreeBSD and macOS arm issues), we'll run it through the buildbots again.

**Comment by erlend-aasland** [comment]
SQLite is now detected without warnings on FreeBSD. It does however still fail bco. `getaddrinfo`. It also fails like that on `main`. 

```
Fatal: You must get working getaddrinfo() function.
       or you can specify "--disable-ipv6".
```

**Comment by erlend-aasland** [comment]
`ARM64 macOS` is also fine now. Nice.

File: configure.ac (hunk 10 above)
Comment (erlend-aasland) [reply configure.ac]: I opened GH-29637 for this. Let me know what you think.

**Review by tiran — COMMENTED** [review]
Tests are passing!

**Review by tiran — APPROVED** [review]

**Comment by erlend-aasland** [comment]
> Tests are passing!

Finally 😆

**Comment by koobs** [comment]
Great job!
